pkg/operator/apiserver/controller/apiservice/apiservice_controller.go

```diff
@@ -90,11 +101,14 @@ func (c *APIServiceController) sync(ctx context.Context, syncCtx factory.SyncCon
 		return nil
 	}
 
-	apiServices, err := c.getAPIServicesToManageFn()
+	enabledApiServices, disabledApiServices, err := c.getAPIServicesToManageFn()
 	if err != nil {
 		return err
 	}
-	ready, err := c.precondition(apiServices)
+
+	disErr := c.syncDisabledAPIServices(ctx, disabledApiServices)
+
+	ready, err := c.preconditionForEnabledAPIServices(enabledApiServices)
```

Comment:
what if both `syncDisabledAPIServices` and `preconditionForEnabledAPIServices` return some error?
do we want to report errors from `syncDisabledAPIServices` ?
i think the errors will be ignored as it is right now.

Reply:
If `syncDisabledAPIServices` returns an error, there are two cases that can happen:
- `preconditionForEnabledAPIServices` returns `err != nil` or `!ready` in which case the `APIServicesAvailable` condition gets reported as false. Making error returned by `syncDisabledAPIServices` kinda obsolete since the `APIServicesAvailable` is already false.
- `preconditionForEnabledAPIServices` returns `err == nil` and `ready` in which case the error returned by `syncDisabledAPIServices` gets checked. Producing `APIServicesAvailable` condition set to false. Just with a different error string.

Reply:
Does setting `APIServicesAvailable` when `syncDisabledAPIServices returns some errors` make sense ?

If all services were available and we failed to remove one service would you expect to find the error in `APIServicesAvailable` ? Maybe we should report it as a separate condition? (Especially that we already made distinction between enabled/disabled on the API level)

Reply:
The current implementation does not make the reported condition worse. Right now it will get reported with `Reason: Error`.  Whether that happens before or after `preconditionForEnabledAPIServices` makes no difference. Failing to remove or create an APIservice object falls under the same category of an error. If we decide to distinguish between error related to removing and creating an APIService, we should probably introduce a new condition for both `Failed to create` and `Failed to check discovery`.

Reply:
I wonder whether there's a value in reporting non-nil error for `return c.syncDisabledAPIServices(ctx, append(enabledApiServices, disabledApiServices...))` under:
```
	switch operatorConfigSpec.ManagementState {
	case operatorsv1.Managed:
	case operatorsv1.Unmanaged:
		return nil
	case operatorsv1.Removed:
		enabledApiServices, disabledApiServices, err := c.getAPIServicesToManageFn()
		if err != nil {
			return err
		}
		return c.syncDisabledAPIServices(ctx, append(enabledApiServices, disabledApiServices...))
```

Reply:
As of today `APIServicesAvailable` was reserved for checking availability which was defined as checked if:
- endpoints were assigned to services
- `apiregistrationv1.Available` condition was set on services
- discovery endpoint works for services

I'm just wondering if it makes sense to report `disabled` services on the same condition.

Reply:
I see what you mean now. A removed API service is no longer part of what is considered as available API services.

From https://github.com/openshift/api/blob/master/operator/v1/types.go#L154-L155:
```
// Degraded indicates that the operator (not the operand) is unable to fulfill the user intent
OperatorStatusTypeDegraded = "Degraded"
```
A new condition should probably be of type `Degraded`. I.e. disabled API services which are still available needs to be removed so the user intent is fulfilled. What about `APIServicesDegraded`?

Reply:
I'm not sure if we want to go `Degraded`. I think that:

```
Type:    "APIServicesDeleted",
Status:  operatorv1.ConditionFalse,
Reason:  "DisabledAPIServicesPresent",
Message: err.Error(),
```

should be fine.

Reply:
Also, We could split the `APIServiceController.sync` method into two flows for simplicity. I think I would do:
```
func (c *APIServiceController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
	operatorConfigSpec, _, _, err := c.operatorClient.GetOperatorState()
	if err != nil {
		return err
	}

	switch operatorConfigSpec.ManagementState {
	case operatorsv1.Managed:
	case operatorsv1.Unmanaged:
		return nil
	case operatorsv1.Removed:
		enabledApiServices, disabledApiServices, err := c.getAPIServicesToManageFn()
		if err != nil {
			return err
		}
		return c.syncDisabledAPIServices(ctx, append(enabledApiServices, disabledApiServices...))
	default:
		syncCtx.Recorder().Warningf("ManagementStateUnknown", "Unrecognized operator management state %q", operatorConfigSpec.ManagementState)
		return nil
	}

       var errors []err
       // or if we rename syncDisabledAPIServices to something else then
       // we could call it syncDisabledAPIServices (which would also set the new status)
       err = c.reconcileDisabledServices()
       if err != nil { // append to errors }
       
       err = c.syncEnabledServices()
       if err != nil { // append to errors }

      return errors
}

Reply:
Would `syncEnabledServices` include calling the `preconditionForEnabledAPIServices` as well?

Reply:
yes, I think so :)

Reply:
as an optimisation step we could collect `conditions` from the individual `sync*` methods and update the config just once :)